src/bridges/chainport.service.ts

```diff
@@ -0,0 +1,345 @@
+        }),
+      ),
+    );
+  }
+
+  private async getMeta(): Promise<{
+    maintenance: boolean;
+    cp_network_ids: Record<string, ChainportNetwork>;
+  }> {
+    const apiurl = this.config.get<string>('CHAINPORT_API_URL');
+    const url = new URL(`/meta`, apiurl);
+
+    const result = await this.makeChainportRequest<{
+      maintenance: boolean;
+      cp_network_ids: Record<string, ChainportNetwork>;
+    }>(url.toString());
+
+    return result.data;
+  }
+
+  async getNetworks(): Promise<ChainportNetwork[]> {
+    const meta = await this.getMeta();
+
+    const validateResult = chainportNetworkArraySchema.validate(
+      Object.values(meta.cp_network_ids),
+      {
+        stripUnknown: true,
+      },
+    );
+
+    if (validateResult.error) {
+      throw new Error(
+        `Invalid Chainport response: ${validateResult.error.message}`,
+      );
+    }
+
+    return validateResult.value;
+  }
+
+  async getVerifiedTokens(): Promise<ChainportToken[]> {
+    const version = this.config.get<number>('CHAINPORT_API_VERSION');
+    const apiurl = this.config.get<string>('CHAINPORT_API_URL');
+    const url = new URL(`/token/list`, apiurl);
+
+    if (version === 1) {
+      url.searchParams.append('network_name', 'IRONFISH');
+    } else {
+      const networkId = this.config.get<string>('CHAINPORT_NETWORK_ID');
+      url.searchParams.append('network_id', networkId);
+    }
+
+    const result = await this.makeChainportRequest<ChainportTokenListResponse>(
+      url.toString(),
+    );
+
+    const validateResult = chainportTokenArraySchema.validate(
+      result.data.verified_tokens,
+      {
+        stripUnknown: true,
+      },
+    );
+
+    if (validateResult.error) {
+      throw new Error(
+        `Invalid Chainport response: ${validateResult.error.message}`,
+      );
+    }
+    return validateResult.value;
+  }
+
+  async getTokenPaths(tokenId: number): Promise<ChainportNetwork[]> {
+    const version = this.config.get<number>('CHAINPORT_API_VERSION');
+    const apiurl = this.config.get<string>('CHAINPORT_API_URL');
+
+    const metaResult = await this.getMeta();
+
+    let networkList: { label: string }[] = [];
+    if (version === 1) {
+      const tokenListUrl = new URL(`/token/list`, apiurl);
+      tokenListUrl.searchParams.append('network_name', 'IRONFISH');
+
+      const tokenListResult = await this.makeChainportRequest<{
+        verified_tokens: { id: number; target_networks: number[] }[];
+      }>(tokenListUrl.toString());
+      const sourceToken = tokenListResult.data.verified_tokens.find(
+        (t) => t.id === tokenId,
+      );
+
+      if (!sourceToken) {
+        throw new NotFoundException();
+      }
+
+      networkList = sourceToken.target_networks.flatMap((n) => {
+        const network = metaResult.cp_network_ids[n.toString()];
+        if (!network) {
+          this.logger.error(
+            `Network ${n} for token ${tokenId} not found in meta`,
+            new Error().stack ?? '',
```

Comment:
Does `new Error().stack` always return some stack we can use? Or should we just use ''`?

Reply:
Yep! It returns the stack at the current location